src/parser/shared/modules/resources/resourcetracker/ResourceTracker.ts

```diff
@@ -7,76 +7,189 @@ import Events, {
+  /** Current resource max */
+  max: number;
+  /** Current rate of continuous resource gain, in resource per second */
+  rate: number;
+  /** Wasted resources caused by the instant change of resources */
+  changeWaste: number;
+  /** Wasted resources since the last update due to recharge */
+  rateWaste: number;
+  /** True iff resources are capped AFTER the change */
+  atCap: boolean;
 };
 
+type ResourceUpdateType =
+  /** A heal or damage event's classResources, or to mark hitting max resources */
+  | 'update'
+  /** Player spent resource, as shown in a cast's classResources */
+  | 'spend'
+  /** Player drained resource, as shown in a DrainEvent */
+  | 'drain'
+  /** Player gained resource, as shown in a ChangeResourceEvent */
+  | 'gain';
+
+const DEBUG = false;
+
 /**
  * This is an 'abstract' implementation of a framework for tracking resource generating/spending.
- * Extend it by following the instructions in the TODO comments below
+ * Extend it by defining the resource and maxResource fields. Other functions can also be
+ * overridden to handle unusual behavior - see comments in class for details.
  */
 class ResourceTracker extends Analyzer {
   static dependencies = {
     eventEmitter: EventEmitter,
+    haste: Haste,
     // Optional dependency for the `resourceCost` prop of events
     // spellResourceCost: SpellResourceCost,
   };
-
   protected eventEmitter!: EventEmitter;
+  protected haste!: Haste;
 
-  current = 0;
-  resourceUpdates: ResourceUpdate[] = [];
+  /////////////////////////////////////////////////////////////////////////////
+  // Overrides - implementer must set these values in constructor!
+  //
 
-  // stores resource gained/spent/wasted by ability ID
-  buildersObj: { [index: number]: BuilderObj } = {};
-  spendersObj: { [index: number]: SpenderObj } = {};
-
-  // TODO set this to the resource you wish to track constructor.. see the appropriate objects in game/RESOURCE_TYPES
+  /** The resource to track */
   resource!: Resource;
-
-  // TODO a classes 'main' resource passes the max along with events, but for other resources this may need to be defined
+  /** The maximum amount of the resource */
   maxResource!: number;
 
+  /** Resource's base regeneration rate, in units per second. Leave as 0 for non-regenerating resources.
+   *  For regenerating resources, override with the value appropriate to the spec/talents.
+   *  Implementer should also modify this value when non-haste things change the rate */
+  baseRegenRate = 0;
+  /** If the resource's regeneration is effected by haste. Override according to your spec */
+  isRegenHasted = true;
+
+  // END override values
+
+  /** Time ordered list of resource updates */
+  resourceUpdates: ResourceUpdate[] = [];
+  /** Tracked builders, indexed by spellId */
+  buildersObj: { [index: number]: BuilderObj } = {};
+  /** Tracked spenders, indexed by spellId */
+  spendersObj: { [index: number]: SpenderObj } = {};
+
   constructor(options: Options) {
```

Comment:
it would be a breaking change to have the constructor set the params that are required, right?

like:

```tsx
type ResourceTrackerParams = Pick<ResourceTracker, 'resource' | 'maxResource'> & Partial<Pick<ResourceTracker, 'baseRegenRate' | 'isRegenHasted'>>;

constructor(options: Options, params: ResourceTrackerParams) {
  this.resource = params.resource;
  this.maxResource = params.maxResource;
  this.baseRegenRate = params.baseRegenRate ?? this.baseRegenRate;
  this.isRegenHasted = params.isRegenHasted ?? this.isRegenHasted;

  // ...the rest of the constructor
}
```

Reply:
It's certainly breaking for all the existing extenders of this class, but wouldn't be too challenging a refactor. I was unsure if I'd have access to `selectedCombatant` when building this way, because for example Feral EnergyTracker needs to check a talent rank to figure out what the baseRegenRate is

Reply:
In fact I think I explicitly had this problem when designing BuffCountGraph - buffSpecs used to be passed in the constructor but it didn't work because I needed dynamic combatantInfo stuff